desc/objectives/_coils.py

```diff
@@ -597,3 +597,207 @@ def compute(self, params, constants=None):
+        """Build constant arrays.
+
+        Parameters
+        ----------
+        use_jit : bool, optional
+            Whether to just-in-time compile the objective and derivatives.
+        verbose : int, optional
+            Level of output.
+
+        """
+        eq = self._eq
+        if self._eval_grid is None:
+            eval_grid = LinearGrid(
+                L=eq.L_grid, M=eq.M_grid, zeta=jnp.array(0.0), NFP=eq.NFP
+            )
+            self._eval_grid = eval_grid
+        eval_grid = self._eval_grid
+
+        warnif(
+            not np.allclose(eval_grid.nodes[:, 2], eval_grid.nodes[0, 2]),
+            UserWarning,
+            "Evaluation grid should be at constant zeta",
+        )
+        if self._normalize:
+            self._normalization = eq.Psi
+
+        # ensure vacuum eq, as is unneeded for finite beta
+        pres = np.max(np.abs(eq.compute("p")["p"]))
+        curr = np.max(np.abs(eq.compute("current")["current"]))
+        errorif(
+            pres > 1e-8,
+            ValueError,
+            f"Pressure is non-zero (max {pres} Pa), "
+            + "this objective is unneeded at finite beta.",
+        )
+        errorif(
+            curr > 1e-8,
+            ValueError,
+            f"Current is non-zero (max {curr} A), "
+            + "this objective is unneeded at finite beta.",
+        )
+
+        # eval_grid.num_nodes for quad flux cost,
+        self._dim_f = 1
+        self._equil_data_keys = ["|e_rho x e_theta|", "R", "phi", "Z"]
+
+        timer = Timer()
+        if verbose > 0:
+            print("Precomputing transforms")
+        timer.start("Precomputing transforms")
+
+        data = eq.compute(
+            ["R", "phi", "Z", "|e_rho x e_theta|", "n_zeta"], grid=eval_grid
+        )
+
+        plasma_coords = jnp.array([data["R"], data["phi"], data["Z"]]).T
+
+        self._constants = {
+            "plasma_coords": plasma_coords,
+            "equil_data": data,
+            "quad_weights": 1.0,
+        }
+
+        timer.stop("Precomputing transforms")
+        if verbose > 1:
+            timer.disp("Precomputing transforms")
+
+        super().build(use_jit=use_jit, verbose=verbose)
+
+    def compute(self, field_params=None, constants=None):
+        """Compute toroidal flux.
+
+        Parameters
+        ----------
+        field_params : dict
+            Dictionary of field degrees of freedom,
+            eg FourierCurrentPotential.params_dict or CoilSet.params_dict
+        constants : dict
+            Dictionary of constant data, eg transforms, profiles etc. Defaults to
+            self.constants
+
+        Returns
+        -------
+        f : float
+            Toroidal flux from coils and external field
+
+        """
+        if constants is None:
+            constants = self.constants
+
+        data = constants["equil_data"]
+        plasma_coords = constants["plasma_coords"]
+
+        B = self._field.compute_magnetic_field(
```

Comment:
I think `field` and the grids should be in `constants`, not referenced via `self`

Reply:
I theory it can avoid the "constant folding" stuff but not sure how much that really helps.